### PR TITLE
initial gnss derived tec decoder 

### LINKED ISSUE #1504: create a total electron content decoder netcdf to ioda
## Description

create a total electron content decoder from netCDF to IODA format

need to confirm if this is a standard file, and how to adapt for other receivers beyond COSMIC-2

## Requirements

python interface to IODA

## Acceptance Criteria (Definition of Done)
a IODA file is output containing the TEC

### PR DESCRIPTION
**Pull request by BenjaminRuston**
build-group=https://github.com/JCSDA-internal/ioda/pull/1256

## Description

This decodes a COSMIC-2 TEC file and puts it into IODA format.

This is related to the IODA PR:
https://github.com/JCSDA-internal/ioda/pull/1256

## Issue(s) addressed

Resolves #1504 

## Dependencies

List the other PRs that this PR is dependent on:
- [x] https://github.com/JCSDA-internal/ioda/pull/1256

## Impact

Initial capability for TEC decoding from netCDF to IODA

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have run the unit tests before creating the PR

### CODE REVIEW

**Comment by BenjaminRuston** [comment]
@davidkuhl this is just a start, but see if I can get a functional decoder soon, could have short cut it but trying to be generally thorough in the creation of the initial prototype

would be good to get other TEC files from potentially other constellations beyond COSMIC-2 and make sure this is the latest format of the files

File: src/gnssro/gnss_tec_bufr2ioda.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+    return meta_data_keys
+
+
+def def_meta_types():
+
+    meta_data_types = {
+        "dateTime": "long",
+        "elevationAngleGNSSlink": "float",
+        "GNSSxECFPosition": "float",
```
Comment (BenjaminRuston) [inline src/gnssro/gnss_tec_bufr2ioda.py]: @davidkuhl I'm coming up with these new IODA entry names, but we'll need to but a bit more formal and likely should ask NOAA SWPC but trying to make this a bit more descriptive and also generic so `x_GPS` now `GNSSxECFPosition` and `x_LEO` now `SATxECFPosition`

**Review by davidkuhl — APPROVED** [review]

**Review by davidkuhl — APPROVED** [review]

File: src/gnssro/gnss_tec_bufr2ioda.py (hunk 1 above)
Comment (davidkuhl) [reply src/gnssro/gnss_tec_bufr2ioda.py]: Let me see if Victoriya has any ideas on more descriptive names for these.

Comment (BenjaminRuston) [reply src/gnssro/gnss_tec_bufr2ioda.py]: @davidkuhl this is not ready yet... ;0)

will let you know when I generate a file, we will also bring in other Victoriya for certain to discuss convention

File: src/gnssro/gnss_tec_bufr2ioda.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            ay_leo = np.concatenate((ay_leo, y_leo), axis=None)
+            az_leo = np.concatenate((az_leo, z_leo), axis=None)
+            ax_gps = np.concatenate((ax_gps, x_gps), axis=None)
+            ay_gps = np.concatenate((ay_gps, y_gps), axis=None)
+            az_gps = np.concatenate((az_gps, z_gps), axis=None)
+            atec = np.concatenate((atec, tec), axis=None)
+            aelev = np.concatenate((aelev, elev), axis=None)
+            aUT_RO = np.concatenate((aUT_RO, UT), axis=None)
+            aRO_ind = np.concatenate((aRO_ind, RO_ind), axis=None)
+            aunique_number = np.concatenate((aunique_number, unique_number), axis=None)
```
Comment (BenjaminRuston) [inline src/gnssro/gnss_tec_bufr2ioda.py]: need to change this to `sequenceNumber`

**Comment by BenjaminRuston** [comment]
adding letter to number follow WMO BUFR:
image

**Review by fcvdb — APPROVED** [review]
The converter was successfully tested on Narwhal and Nautilus with file `podTc2_C2E3.2020.275.08.18.0013.G10.01_0001.0001_nc`

**Comment by BenjaminRuston** [comment]
@davidkuhl so I've added a function to compute the semi-accurate lat/lon from the ECF coordinates.  Think this will be helpful as now the IODA file will contain the same `required` data as the other data types.  The operator is fully welcome to not use these and proceed to use the ECF coordinates.  

I've started the IODA PR too, added to this PR description think have a good proposal to go forward with

File: src/gnssro/gnss_tec_bufr2ioda.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+float_missing_value = iconv.get_default_fill_val(np.float32)
+int_missing_value = iconv.get_default_fill_val(np.int32)
+long_missing_value = iconv.get_default_fill_val(np.int64)
+
+locationKeyList = [
+    ('elevationAngleGNSS', 'float'),
+    ('latitude', 'float'),
+    ('longitude', 'float'),
+    ('xECFPosition', 'float'),
```
Comment (BenjaminRuston) [inline src/gnssro/gnss_tec_bufr2ioda.py]:
```suggestion
    ('xECEFPosition', 'float'),
```

when I look up this coordinate system `ECEF` is more commonly used, should we adopt this as the standard?

File: src/gnssro/gnss_tec_bufr2ioda.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+def xyz2llh(x, y, z):
+    """
+    Function to convert xyz ECEF to llh
+    convert cartesian coordinate into geographic coordinate
+    ellipsoid definition: WGS84
+      a = 6,378,137m
+      f = 1 / 298.257
+
+    Input
+      x: coordinate X meters
```
Comment (BenjaminRuston) [inline src/gnssro/gnss_tec_bufr2ioda.py]:
```suggestion
      x: coordinate x meters
```

**Comment by BenjaminRuston** [comment]
@davidkuhl and Victoriya the xyz2llh was abstracted from here:
https://gis.stackexchange.com/questions/265909/converting-from-ecef-to-geodetic-coordinates

**Comment by BenjaminRuston** [comment]
these data will use the international reference ionosphere and an interface will develop with this version:
https://github.com/JCSDA-internal/pyiri-jedi

File: src/gnssro/gnss_tec_bufr2ioda.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+locationKeyList = [
+    ('elevationAngleGNSS', 'float'),
+    ('latitude', 'float'),
+    ('longitude', 'float'),
+    ('xECEFPosition', 'float'),
+    ('yECFPosition', 'float'),
+    ('zECFPosition', 'float'),
+    ('xECFPositionGNSS', 'float'),
+    ('yECFPositionGNSS', 'float'),
+    ('zECFPositionGNSS', 'float'),
```
Comment (BenjaminRuston) [inline src/gnssro/gnss_tec_bufr2ioda.py]:
```suggestion
    ('yECEFPosition', 'float'),
    ('zECEFPosition', 'float'),
    ('xECEFPositionGNSS', 'float'),
    ('yECEFPositionGNSS', 'float'),
    ('zECEFPositionGNSS', 'float'),
```

**Review by huishao-r — COMMENTED** [review]
It seems good to me. Based on feedback from SWPC, there might be some quality flag from the data file?  Should we check the raw data and see whether there is any QC related variable not yet included?

**Pull request merged**